storage/innobase/fsp/fsp0fsp.cc

```diff
@@ -3196,6 +3200,68 @@ class fsp_xdes_old_page
   }
 };
 
+/** Used during freeing of unused segment in system tablespace.
+Stores the "to be modified" header, inode pages, extent descriptor
+page and its old page state */
+struct fsp_old_inode
+{
+  std::map<uint32_t, buf_block_t*> m_old_inode_entry;
+public:
+  dberr_t insert(uint32_t page_no, mtr_t *mtr)
+  {
+    if (m_old_inode_entry[page_no])
+      return DB_SUCCESS;
+    dberr_t err= DB_SUCCESS;
+    buf_block_t *block=
+      buf_page_get_gen(page_id_t{0, page_no}, 0, RW_SX_LATCH,
+                       nullptr, BUF_GET, mtr, &err);
+    if (!block)
+      return err;
+    buf_block_t *old= buf_LRU_get_free_block(have_no_mutex_soft);
+    if (!old) return DB_OUT_OF_MEMORY;
+
+    memcpy_aligned<UNIV_PAGE_SIZE_MIN>(
+      old->page.frame, block->page.frame, srv_page_size);
+    m_old_inode_entry[page_no]= old;
```

Comment:
At the start of this function, the `operator[]` call already inserted a `nullptr` to the `std::map`. We could have saved a reference to that, e.g.,
```c++
buf_block_t *&old= m_old_inode_entry[page_no];
if (old)
  return DB_SUCCESS;
//…
old= buf_LRU_get_free_block(have_no_mutex_soft);
return old ? DB_OUT_OF_MEMORY : DB_SUCCESS;
```
However, I do not quite understand the need for storing a copy of the page. Logic like this would make sense when we are shrinking the system tablespace and we expect to free most pages, rewriting only a few surviving pages of the system tablespace. In that case we wanted to minimize the amount of redo log written, so that everything can be done in a single atomic mini-transaction, with at most 2 MiB of log records.

Here we are handling a special case and we are allowed to use as many mini-transactions as it takes.

Can we structure the mini-transactions in such a way that we free only one leaked page per mini-transaction, after performing all necessary consistency checks? In that way, there should be no need to allocate any shadow pages. Yes, we may be modifying the same pages over and over again, but the buffer pool will cache the write-backs. Yes, we may write a lot of log records, but this is a very rarely executed clean-up of corrupted data, and it will not be executed by default.